blog/content/post/2023-06-07-twpm-spi-fix.md

```diff
@@ -0,0 +1,855 @@
+is a significant challenge, especially when operating as a slave. TPM-specific
+features complicate things further.
+
+## Limitations of STM32L476
+
+STM32L476 has SPI capable of frequencies up to a (theoretical) limit of 40 MHz,
+which is half of the maximum clock that can be provided to Cortex-M and AHB/APB
+buses. Other limiting factors include maximum GPIO speed, DMA transfer speed,
+and performance of the firmware itself.
+
+## Creating SPI slave on Zephyr
+
+Zephyr is our platform of choice primarily due to its portability (we will
+target non-STM32 platforms too). I will briefly describe the outcome of my early
+tests done on Zephyr and why it was terrible.
+
+The application just transmits a static sequence of bytes:
+
+```c
+const struct device *spi_dev = NULL;
+struct spi_config spi_cfg = {
+	.frequency = 25000000,
+	.operation = SPI_WORD_SET(8) | SPI_TRANSFER_MSB | SPI_OP_MODE_SLAVE,
+};
+
+const uint8_t testdata[32] = {
+	0x3d, 0x7f, 0x85, 0xc3, 0x86, 0x2f, 0x14, 0x2a, 0xa2, 0x67, 0x1d, 0xd7, 0xfa, 0xa8, 0x3a, 0x42,
+	0xf8, 0x12, 0xd2, 0xa1, 0x04, 0xcc, 0xe2, 0xc6, 0x78, 0x73, 0x09, 0xe6, 0xd8, 0xc5, 0x0e, 0xba,
+};
+
+void main(void) {
+	spi_dev = device_get_binding(DEVICE_DT_NAME(DT_NODELABEL(spi2)));
+	if (!device_is_ready(spi_dev)) {
+		printk("SPI controller is not ready!\n");
+		return;
+	}
+
+	struct spi_buf spi_tx_buf = {
+		.buf = (void *)testdata,
+		.len = sizeof testdata,
+	};
+
+	struct spi_buf_set spi_tx_buf_set = {
+		.buffers = &spi_tx_buf,
+		.count = 1,
+	};
+
+	while (true) {
+		int ret = spi_write(spi_dev, &spi_cfg, &spi_tx_buf_set);
+		if (ret < 0)
+			printk("spi_write failed: %d\n", ret);
+		else
+			printk("spi transfer complete\n");
+	}
+}
+```
+
+It is necessary to enable some configs:
+
+```shell
+CONFIG_GPIO=y
+CONFIG_SPI=y
+CONFIG_SPI_SLAVE=y
+CONFIG_SPI_STM32_DMA=y
+```
+
+And configure DMA channels:
+
+```shell
+&dma1 {
+	status = "okay";
+};
+
+&spi1 {
+	status = "disabled";
+};
+
+&spi2 {
+	/*
+	 * See https://docs.zephyrproject.org/3.0.0/reference/devicetree/bindings/dma/st%2Cstm32-dma-v2.html
+	 */
+	dmas = <&dma1 5 1 0x20440>,
+	       <&dma1 4 1 0x20480>;
+	dma-names = "tx", "rx";
+};
+
+&spi3 {
+	status = "disabled";
+};
+```
+
+If we run this code we will see SPI errors running down the console:
+
+```shell
+spi_write failed: -11
+spi_write failed: -11
+spi_write failed: -11
+spi_write failed: -11
+spi_write failed: -11
+```
+
+This is the first problem with Zephyr's SPI driver - each transfer has a
+one-second timeout. While this may be desirable behavior for SPI master (it
+could be used for error recovery, for example, to power cycle the slave if it
+doesn't respond), it breaks SPI slave. The slave must be ready to give a
+response when the transfer commences - appropriate data must already be loaded
+in FIFO. Here we get stuck in an endless loop, queuing the transfer, aborting
+it, and queuing it again.
+
+The problem can be worked around by patching the `wait_dma_rx_tx_done` function
```

Comment:
Can we link to code in github for these functions?

Reply:
Done